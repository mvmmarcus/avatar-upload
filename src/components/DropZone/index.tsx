import React from "react";

import Avatar from "components/Avatar";
import Button from "components/Button";
import Slider from "components/Slider";

import {
  Content,
  ErrorMessage,
  Header,
  Subtitle,
  Title,
  TitleIcon,
} from "./styles";

type DropZoneProps = {
  avatarUrl?: string | ArrayBuffer | null;
  errorMessage?: string;
  isCropping?: boolean;
  onFinishCrop?: () => void;
  onTryAgain?: () => void;
  onScaleChange?: (scale: number) => void;
};

const DropZone = ({
  avatarUrl,
  errorMessage,
  isCropping = false,
  onTryAgain,
  onFinishCrop,
  onScaleChange,
}: DropZoneProps) => {
  const handleFinishCrop = () => {
    !!onFinishCrop && onFinishCrop();
    !!onScaleChange && onScaleChange(1);
  };

  const handleTryAgain = () => {
    !!onTryAgain && onTryAgain();
  };

  const handleScaleChange = (scale: number) => {
    !!onScaleChange && onScaleChange(scale);
  };

  if (errorMessage) {
    return (
      <Content hasAvatar>
        {!!avatarUrl && <Avatar urlImg={avatarUrl as string} />}
        <ErrorMessage>{errorMessage}</ErrorMessage>
        <Button variant="link" onClick={handleTryAgain}>
          Try again
        </Button>
      </Content>
    );
  }

  return (
    <>
      {isCropping ? (
        <Content hasAvatar={!!avatarUrl}>
          <Slider onSave={handleFinishCrop} onValueChange={handleScaleChange} />
        </Content>
      ) : (
        <Content
          hasAvatar={!!avatarUrl}
          isSuccess={!errorMessage && !isCropping}
        >
          <Header>
            <TitleIcon src="/img/media.svg" alt="media icon" />
            <Title>Organization Logo</Title>
          </Header>
          <Subtitle>Drop the image here or click to browse.</Subtitle>
        </Content>
      )}
    </>
  );
};

export default DropZone;
