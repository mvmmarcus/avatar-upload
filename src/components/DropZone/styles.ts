import styled, { css, DefaultTheme } from "styled-components";
import media from "styled-media-query";

type ContentProps = {
  hasAvatar: boolean;
  isSuccess?: boolean;
};

const contentModifiers = {
  hasAvatar: (theme: DefaultTheme, isSuccess?: boolean) => css`
    align-items: ${isSuccess ? "center" : "flex-start"};
    margin-left: ${theme.spacings.xsmall};

    ${media.greaterThan("small")`
      margin-left: ${theme.spacings.medium};
    `}
  `,
};

export const Content = styled.div<ContentProps>`
  ${({ theme, hasAvatar, isSuccess = false }) => css`
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;

    ${hasAvatar && contentModifiers.hasAvatar(theme, isSuccess)}
  `}
`;

export const Header = styled.header`
  display: flex;
  align-items: center;
  pointer-events: none;
  pointer-events: none;
`;

export const TitleIcon = styled.img`
  ${({ theme }) => css`
    margin-right: ${theme.font.sizes.xsmall};
  `}
`;

export const Title = styled.p`
  ${({ theme }) => css`
    line-height: 2.5rem;
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.medium};
    color: ${theme.colors.gray06};
  `}
`;

export const Subtitle = styled.p`
  ${({ theme }) => css`
    line-height: 2.5rem;
    pointer-events: none;
    text-align: center;
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.normal};
    color: ${theme.colors.gray05};
  `}
`;

export const ErrorMessage = styled.p`
  ${({ theme }) => css`
    pointer-events: none;
    line-height: ${theme.font.sizes.xxlarge};
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.normal};
    color: ${theme.colors.red};
  `}
`;
