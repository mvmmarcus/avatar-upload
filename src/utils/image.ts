type CropImageProps = {
  imageUrl: string;
  cropWidth?: number;
  preserveAspectRatio?: boolean;
  previewImageRatio?: number;
};

const addImageProcess = async (
  imageSource: string
): Promise<HTMLImageElement> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = reject;
    img.src = imageSource;
  });
};

export const cropImage = async ({
  imageUrl,
  cropWidth = 300,
  preserveAspectRatio = true,
  previewImageRatio = 1,
}: CropImageProps) => {
  const image = await addImageProcess(imageUrl);

  const ratio = image.height / image.width;

  const canvas = document.createElement("canvas");
  const context = canvas.getContext("2d");

  if (preserveAspectRatio) {
    canvas.width = image.width < cropWidth ? image.width : cropWidth;
    canvas.height = image.width < cropWidth ? image.height : cropWidth * ratio;

    context?.drawImage(image, 0, 0, canvas.width, canvas.height);
  } else {
    canvas.width = previewImageRatio * image.width;
    canvas.height = canvas.width;

    const cropDistanceX = (image.width - canvas.width) / 2;
    const cropDistanceY = (image.height - canvas.height) / 2;

    context?.drawImage(
      image,
      cropDistanceX,
      cropDistanceY,
      canvas.width,
      canvas.height,
      0,
      0,
      canvas.width,
      canvas.height
    );
  }

  const croppedImageUrl = canvas.toDataURL();

  return croppedImageUrl;
};

export const uploadImage = (file: File): Promise<ProgressEvent<FileReader>> => {
  return new Promise((resolve, reject) => {
    const fileReader = new FileReader();

    fileReader.onload = (event) => resolve(event);
    fileReader.onerror = reject;
    fileReader.readAsDataURL(file);
  });
};
