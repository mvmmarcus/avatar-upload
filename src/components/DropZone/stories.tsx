import React from "react";

import { Story, Meta } from "@storybook/react/types-6-0";

import DropZone, { DropZoneProps } from ".";

export default {
  title: "DropZone",
  component: DropZone,
  argTypes: {
    avatarUrl: {
      control: { disable: true },
    },
    errorMessage: {
      type: "string",
    },
    isCropping: {
      type: "boolean",
    },
    onTryAgain: { action: "Try again" },
    onFinishCrop: { action: "Finish crop" },
  },
} as Meta<DropZoneProps>;

export const Default: Story<DropZoneProps> = (args) => <DropZone {...args} />;

Default.args = {};

export const Cropping: Story<DropZoneProps> = (args) => <DropZone {...args} />;

Cropping.args = {
  isCropping: true,
};

export const Error: Story<DropZoneProps> = (args) => <DropZone {...args} />;

Error.args = {
  errorMessage: "Sorry, the upload failed.",
};
