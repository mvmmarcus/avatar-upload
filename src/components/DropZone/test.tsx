import React from "react";

import { fireEvent, screen } from "@testing-library/react";

import DropZone from ".";
import { renderWithTheme } from "utils/test-utils";

describe("<Slider />", () => {
  it("should render DropZone as default", () => {
    const { container } = renderWithTheme(<DropZone />);

    expect(
      screen.getByRole("img", { name: /media icon/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: /Organization Logo/i })
    ).toBeInTheDocument();
    expect(
      screen.getByText("Drop the image here or click to browse.")
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should change image scale and finish croppping", () => {
    const onFinishCrop = jest.fn();
    const onScaleChange = jest.fn();

    const { container } = renderWithTheme(
      <DropZone
        onFinishCrop={onFinishCrop}
        onScaleChange={onScaleChange}
        isCropping
      />
    );

    const sliderInput = screen.getByTestId("slider-input");
    const saveButton = screen.getByRole("button", { name: /save/i });

    fireEvent.change(sliderInput, {
      target: {
        value: 1.5,
      },
    });
    fireEvent.click(saveButton);

    expect(onScaleChange).toHaveBeenCalledWith(1.5);
    expect(onFinishCrop).toHaveBeenCalledWith();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render with error and click on try again button", () => {
    const onFinishCrop = jest.fn();

    const { container } = renderWithTheme(
      <DropZone
        onTryAgain={onFinishCrop}
        errorMessage="Sorry, the upload failed."
      />
    );

    expect(screen.getByText("Sorry, the upload failed.")).toBeInTheDocument();
    const tryAgainButton = screen.getByRole("button", { name: /try again/i });

    fireEvent.click(tryAgainButton);

    expect(onFinishCrop).toHaveBeenCalledWith();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render with success", () => {
    const unsplashTestImageUrl =
      "https://images.unsplash.com/photo-1645300188328-6708bd739a81?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=688&q=80";

    const { container } = renderWithTheme(
      <DropZone avatarUrl={unsplashTestImageUrl} />
    );

    expect(container.firstChild).toHaveStyle("align-items: center");
    expect(
      screen.getByRole("img", { name: /media icon/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: /Organization Logo/i })
    ).toBeInTheDocument();
    expect(
      screen.getByText("Drop the image here or click to browse.")
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
