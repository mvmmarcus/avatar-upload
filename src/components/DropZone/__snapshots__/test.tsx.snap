// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider /> should change image scale and finish croppping 1`] = `
.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  border: none;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  height: 3.2rem;
  font-size: 1.4rem;
  font-weight: 500;
  background-color: #3D485F;
  border: 0.8rem;
  padding: 0.4rem 3.2rem;
  border-radius: 1.6rem;
}

.c5:hover {
  background-color: #677489;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  max-width: 27.6rem;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 27.6rem;
  margin-bottom: 3.2rem;
}

.c3 {
  line-height: 2.8rem;
  font-size: 1.6rem;
  font-weight: 400;
  color: #677489;
  margin-bottom: 1.2rem;
}

.c4 {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  outline: 0;
  height: 0.2rem;
  min-height: 0.2rem;
  border-radius: 0.1rem;
  background: linear-gradient( to right, #3F80FF 0%, #3F80FF 50%, #B9D1FF 50%, #B9D1FF 100% );
}

.c4::-webkit-slider-thumb {
  -webkit-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

.c4::-moz-range-thumb {
  -moz-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <span
        class="c3"
      >
        Crop
      </span>
      <input
        class="c4"
        data-testid="slider-input"
        max="2"
        min="1"
        step=".1"
        type="range"
        value="1.5"
      />
    </div>
    <button
      class="c5"
    >
      Save
    </button>
  </div>
</div>
`;

exports[`<Slider /> should render DropZone as default 1`] = `
.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  pointer-events: none;
  pointer-events: none;
}

.c2 {
  margin-right: 1.2rem;
}

.c3 {
  line-height: 2.5rem;
  font-size: 1.4rem;
  font-weight: 500;
  color: #495567;
}

.c4 {
  line-height: 2.5rem;
  pointer-events: none;
  text-align: center;
  font-size: 1.4rem;
  font-weight: 400;
  color: #677489;
}

<div
  class="c0"
>
  <header
    class="c1"
  >
    <img
      alt="media icon"
      class="c2"
      src="/img/media.svg"
    />
    <h3
      class="c3"
    >
      Organization Logo
    </h3>
  </header>
  <p
    class="c4"
  >
    Drop the image here or click to browse.
  </p>
</div>
`;

exports[
  `<Slider /> should render with error and click on try again button 1`
] = `
.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  border: none;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  height: 3.2rem;
  font-size: 1.4rem;
  font-weight: 500;
  background-color: #3D485F;
  border: 0.8rem;
  padding: 0.4rem 3.2rem;
  border-radius: 1.6rem;
  background: transparent;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  padding: 0;
  font-size: 1.6rem;
  line-height: 2.8rem;
  font-weight: 500;
  color: #3D485F;
}

.c2:hover {
  background-color: #677489;
}

.c2:hover {
  background-color: transparent;
  color: #677489;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  margin-left: 1.6rem;
}

.c1 {
  pointer-events: none;
  line-height: 2.8rem;
  font-size: 1.6rem;
  font-weight: 400;
  color: #C64D32;
}

@media (min-width:450px) {
  .c0 {
    margin-left: 3.2rem;
  }
}

<div
  class="c0"
>
  <p
    class="c1"
  >
    Sorry, the upload failed.
  </p>
  <button
    class="c2"
  >
    Try again
  </button>
</div>
`;

exports[`<Slider /> should render with success 1`] = `
.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 1.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  pointer-events: none;
  pointer-events: none;
}

.c2 {
  margin-right: 1.2rem;
}

.c3 {
  line-height: 2.5rem;
  font-size: 1.4rem;
  font-weight: 500;
  color: #495567;
}

.c4 {
  line-height: 2.5rem;
  pointer-events: none;
  text-align: center;
  font-size: 1.4rem;
  font-weight: 400;
  color: #677489;
}

@media (min-width:450px) {
  .c0 {
    margin-left: 3.2rem;
  }
}

<div
  class="c0"
>
  <header
    class="c1"
  >
    <img
      alt="media icon"
      class="c2"
      src="/img/media.svg"
    />
    <h3
      class="c3"
    >
      Organization Logo
    </h3>
  </header>
  <p
    class="c4"
  >
    Drop the image here or click to browse.
  </p>
</div>
`;
